### ISSUE TITLE
Matplotlib/pylab animation doesn't work in 2.0 (it worked in 1.1.8)

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#459]
_From [maxims...@gmail.com](https://code.google.com/u/112498364678121572766/) on 2010-11-20T09:42:35Z_

What steps will reproduce the problem?

 1. Create simple script that uses pylab for animation 
2. Run it in 2.0rc1 What is the expected output? What do you see instead? 1.1.8 (using Run in interactive console): each call to show() updates figure and does not block, script is running -> I get animation
2.0rc1: (Run in interactive console is gone): same script is executed using F5, it wors, show() calls doesn't block BUT the window with graph stays empty. So animation doesn't work.
After script is interrupted (using Ctrl-C), window with figure is updated and I can see my graph. What version of the product are you using? On what operating system? Using 2.0rc1 on latest python(x,y) installation on Windwos XP SP3 32bit.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=459_

**Comment by spyder-issues-migrator** [comment]
_From [maxims...@gmail.com](https://code.google.com/u/112498364678121572766/) on 2010-11-20T07:08:16Z_

Simple example script:

from pylab import *
from time import *
x = arange(5)
y = arange(5)
ion()
hold(False)
plot(x,y)
z = array(transpose(matrix(y))*matrix(x))
p = imshow(z)
show()
for i in xrange(5):
    z = z + 2    
    p.set_data(z)
    clim()
    \#draw()
    print "step"
    sleep(1)

Strangely, draw() doesn't produce animation, but clim() does. This is "killer feature" of spyder for me and I would very much appreciate if you could keep in in 2.0

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-21T06:29:32Z_

Thanks for the sample script, that is the most effective way for me to be able to reproduce the bug you are reporting.

That being said, I have a bad feeling about your script as there is no animation when trying to run it outside Spyder:
  * in a standard Python interpreter: python ani.py (where ani.py contains the exact code above)
  * or in an IPython (-pylab) session

My first guess would be that maybe it was working in Spyder v1.x's interactive console but it was not supposed to!

In other words, when a script does not work outside Spyder, it's not supposed to work in it neither.

**Comment by spyder-issues-migrator** [comment]
_From [maxims...@gmail.com](https://code.google.com/u/112498364678121572766/) on 2010-11-21T06:35:50Z_

Yes, it doesn't work outside Spyder 1.x. That's what makes Spyder so special :) I think easy animation is extremely useful feature, and it will be very good if Spyder 2.x could keep it. After all Spyder is MATLAB-like, and matlab allows animation. Main problem with animation in matplotlib is need for external event loop, and Spyder 1.x seemed to provide one. If some design changes breaks this feature in Spyder 2.x maybe it is possible to keep it as a plugin?

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-21T06:59:58Z_

After reading carefuly your script, here is what you can do to make it work.

Either you chose another backend than Qt:
\----------------------------------------

import matplotlib
matplotlib.use('TkAgg')

from pylab import *
from time import *
x = arange(5)
y = arange(5)
ion()
hold(False)
plot(x,y)
z = array(transpose(matrix(y))*matrix(x))
p = imshow(z)
for i in xrange(5):
    z = z + 2    
    p.set_data(z)
    clim()
    draw()
    print "step"
    sleep(1)


Or you may stay with Qt: (but you have to give it the opportunity to cycle the event loop)
\-----------------------

from PyQt4.QtGui import QApplication
import matplotlib

from pylab import *
from time import *
x = arange(5)
y = arange(5)
ion()
hold(False)
plot(x,y)
z = array(transpose(matrix(y))*matrix(x))
p = imshow(z)
for i in xrange(5):
    z = z + 2    
    p.set_data(z)
    clim()
    draw()
    print "step"
    QApplication.processEvents()
    sleep(1)

I think that matplotlib developers may have forgot to do 'QApplication.processEvents()' in the 'draw' function.

**Comment by spyder-issues-migrator** [comment]
_From [maxims...@gmail.com](https://code.google.com/u/112498364678121572766/) on 2010-11-21T07:11:26Z_

Thanks, that helps a lot! But: first script doesn't work in Spyder 2.x, because Spyder sets matplotlib backend to Qt and doesn't allow it to be changed. But the important part is that second script works perfectly in and outside of Spyder. I will file a bug in matplotlib for QApplication.processEvents() in Qt backend. So I guess you should close this bug.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-21T07:39:51Z_

The first script works with Spyder v2.x: you *can* change Matplotlib's backend in Preferences>Console>External modules (or you may even disable the patch to Matplotlib, thus allowing you to choose the backend you want in your script).

As for the matplotlib's "bug", here is a mail I've just sent: http://old.nabble.com/Simple-animation-with-Qt4-backend-td30272035.html

**Status:** Done

**Comment by spyder-issues-migrator** [comment]
_From [sivananda.redteam@gmail.com](https://code.google.com/u/sivananda.redteam@gmail.com/) on 2014-02-25T05:01:27Z_

Thanks a lot. After a lot of struggle i found solution here.
